proposals/interface-types/working-notes/Types.md

```diff
@@ -37,36 +62,158 @@ Integer values.
+#### String Type
 
-## `record`
+The `string` type denotes a sequence of
+[unicode code points](https://www.unicode.org/glossary/#code_point).
 
-The `record` interface value type is a compound type defined by the following abstract syntax:
 
 ```
-#record ::= record $fields: (field #interface-valtype)*
+stringtype ::= string
 ```
 
-A `record` value is an ordered set of `|$fields|` interface values, where each interface value is of the corresponding type in `$fields`. This is sometimes known as a struct or product type.
+>Note: In a future revision, this may be reinterpreted as a macro of a more
+>explicit designation of the `sequence` type.
+
+Note that there is no implied representation or encoding implied in this.
+
 
-A `record` type is valid iff:
- * `|$fields| > 0`
- * For every `field` in `$fields`
-  * `#interface-valtype` is valid
+### Structured Value Types
 
-## `array`
+Structured types allow the representation of more complex combinations of
+values. There are five forms of structured type: arrays, sequences, records,
+protocols and algebraic variants.
 
-The `array` interface value type is a compound type defined by the following abstract syntax:
+```
+structuretype ::= optiontype | eithertype |
+   arraytype | sequencetype | recordtype | varianttype
+```
+
+#### Option Type
+
+Option types are used to model nullability in types.
 
 ```
-#array ::= array #interface-valtype
+optiontype ::= option type
 ```
 
-An `array` value is a homogenous sequence of interface values with type `$interface-value-type`.
+The `option` type can be viewed as a use of type variants -- with one variant
+being the enumerated symbol `$none` and the other being the single value wrapped
+in a `$some` variant.
+
+#### Either Type
+
+The `either` type is used to model situations where one of two values may be
+returned by a function. A classic case for this is in modeling error returns in
+APIs.
+
+```
+eithertype ::= either type
+```
+
+#### Arrays
+
+An `array` consists of a finite vector of elements, each of which has the same
+type.
+
+```
+arraytype ::= array type
+```
+
+Both the type and the representation of each element of the array is assumed to
+be identical. Furthermore, arrays are assumed to have a know element
+count. (Indexing into arrays is not part of this specification.)
+
+#### Sequences
+
+A sequence consists of an ordered collection of elements; each of which has the
+same type.
+
+Sequence elements may not have identical representation; nor is it necessarily
+known how many elements there are in a sequence.
+
+```
+sequencetype ::= sequence type
+```
+
+#### Records
+
+A `record` type is a tuple composition of types. 
+
+```
+#record ::= record [ vec(interfacetype) ]
+```
+
+Fields in a record are accessed via their index; for convenience, field indices
+may be given a local name:
+
+```
+record [ $name string $age u64]
+```
+
+#### Variant Types
+
+A `variant` consists of an ordered collection of alternative vectors of types.
```

Comment:
Is it necessary to have each alternative be a vector of types if we have record types, which may express vectors of types?

Reply:
It its a vector, the integration with enumerations is much smoother.